v2/pkg/clusterresources/clusterresources.go

```diff
@@ -97,6 +107,152 @@ func (o *ClusterResourcesGenerator) IDMSGenerator(allRelatedImages []v1alpha3.Co
+		if strings.Contains(catalogRef, op.Catalog) {
+			return op.TargetCatalogSourceTemplate
+		}
+	}
+	return ""
+}
+
+func (o *ClusterResourcesGenerator) generateCatalogSource(catalogRef string, catalogSourceTemplateFile string) error {
+
+	catalogSpec, err := image.ParseRef(catalogRef)
+	if err != nil {
+		return err
+	}
+
+	csSuffix := "0" // default value
+	if catalogSpec.IsImageByDigest() {
+		csSuffix = catalogSpec.Digest[:hashTruncLen]
+	} else {
+		tag := catalogSpec.Tag
+		csSuffix = strings.Map(toRFC1035, tag)
+	}
+
+	pathComponents := strings.Split(catalogSpec.PathComponent, "/")
+	catalogRepository := pathComponents[len(pathComponents)-1]
+	catalogSourceName := "cs-" + catalogRepository + "-" + csSuffix
+	errs := validation.IsDNS1035Label(catalogSourceName)
+	if len(errs) != 0 && isValidRFC1123(catalogSourceName) {
+		return fmt.Errorf("error creating catalog source name: %s", strings.Join(errs, ", "))
+	}
+
+	var obj ofv1alpha1.CatalogSource
+	generateWithoutTemplate := false
+	if catalogSourceTemplateFile != "" {
+		obj, err = catalogSourceContentFromTemplate(catalogSourceTemplateFile, catalogSourceName, catalogSpec.Reference)
+		if err != nil {
+			generateWithoutTemplate = true
+			o.Log.Error("error generating catalog source from template. Fall back to generating catalog source without template: %v", err)
+		}
+	}
+	if generateWithoutTemplate || catalogSourceTemplateFile == "" {
+		obj = ofv1alpha1.CatalogSource{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: ofv1alpha1.GroupName + "/" + ofv1alpha1.GroupVersion,
+				Kind:       "CatalogSource",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      catalogSourceName,
+				Namespace: "openshift-marketplace",
+			},
+			Spec: ofv1alpha1.CatalogSourceSpec{
+				SourceType: "grpc",
+				Image:      catalogSpec.Reference,
+			},
+		}
+	}
+	bytes, err := yaml.Marshal(obj)
+	if err != nil {
+		return fmt.Errorf("unable to marshal CatalogSource yaml: %v", err)
+	}
+
+	csFileName := filepath.Join(o.WorkingDir, clusterResourcesDir, catalogSourceName+".yaml")
+	// save IDMS struct to file
+	if _, err := os.Stat(csFileName); errors.Is(err, os.ErrNotExist) {
+		o.Log.Debug("%s does not exist, creating it", csFileName)
+		err := os.MkdirAll(filepath.Dir(csFileName), 0755)
+		if err != nil {
+			return err
+		}
+		o.Log.Debug("%s dir created", filepath.Dir(csFileName))
+	}
+	csFile, err := os.Create(csFileName)
+	if err != nil {
+		return err
+	}
+
+	defer csFile.Close()
+
+	_, err = csFile.Write(bytes)
+	o.Log.Info("%s file created", csFileName)
+
+	return err
+}
+
+func catalogSourceContentFromTemplate(templateFile, catalogSourceName, image string) (ofv1alpha1.CatalogSource, error) {
+	// Initializing catalogSource `obj` from template
+	var obj ofv1alpha1.CatalogSource
+	_, err := os.Stat(templateFile)
+	if os.IsNotExist(err) {
+		return obj, fmt.Errorf("unable to marshal CatalogSource yaml: targetCatalogSourceTemplate does not exist: %s", templateFile)
```

Comment:
"the unable to marshal Catalog yaml" is a bit confusing , could you add a more descriptive pre-fix (or just get rid of it)

Reply:
Done! thanks